[GH-241] Add support for nftables.

a7b81a9086 opened the pull request:
# Description
* add resources `nftables` and `nftables_rule`
* add `libraries/helpers_nftables` to build rules for nftables
* add documentation

## Issues Resolved
closes GH-241

## Check List

- [x] A summary of changes made is included in the CHANGELOG under `## Unreleased`
- [x] New functionality includes testing.
- [x] New functionality has been documented in the README if applicable.

damacus commented on kitchen.yml:
```diff
@@ -92,6 +92,28 @@ suites:
         ubuntu_iptables: true
         redhat7_iptables: true
 
+  - name: nftables
```
You might want to just do an includes here.

a7b81a9086 replied:
Thanks for the hint.

damacus commented on libraries/provider_firewall_nftables.rb:
```diff
@@ -0,0 +1,157 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+class Chef
```
If we're adding a new resource, we should be adding it via the custom resource pattern rather than the old LWRP way.

a7b81a9086 replied:
Ok, I did something, but I am not sure if I did it correctly.  So basically I copied `libraries/provider_firewall_nftables.rb` over to `resource/firewall_nftables.rb` and added some properties, that are already defined as attributes in `libraries/resource_firewall.rb`.

bmhughes commented on attributes/nftables.rb:
```diff
@@ -0,0 +1,22 @@
+default['firewall']['nftables']['defaults'][:policy] = {
```
Let's not inter-mix `String` and `Symbol` keys despite the fact that the `Mash` makes it work either way

bmhughes commented on attributes/nftables.rb:
```diff
@@ -0,0 +1,22 @@
+  forward: 'drop',
+  output: 'accept',
+}
+default['firewall']['nftables']['defaults'][:ruleset] = {
```
Let's not inter-mix `String` and `Symbol` keys despite the fact that the `Mash` makes it work either way

Xorima replied:
I thought this would fail cookstyle, @tas50 ?

a7b81a9086 replied:
FYI: I ran cookstyle 7.8.0 (RuboCop 1.10.0) and it did not complain. I will use strings instead of symbols.

bmhughes commented on libraries/helpers_nftables.rb:
```diff
@@ -0,0 +1,112 @@
+
+        if rule_resource.source
+          source_with_mask = ip_with_mask(rule_resource, rule_resource.source)
+          Chef::Log.fatal source_with_mask
```
Is this definetly the right log level? Seems strange to use `fatal` if it's not conditional

Xorima replied:
Also needs more context around it, if I see this in the log I don't get why it is fatal, what did I do wrong?

a7b81a9086 replied:
I did some debugging there and forgot to remove that line.  There should be no log entry at all, I will fix this.

bmhughes commented on libraries/helpers_nftables.rb:
```diff
@@ -0,0 +1,112 @@
+      end
+
+      def ensure_default_rules_exist(current_node, new_resource)
+        input = new_resource.rules
```
`new_resource.rules || {}` then drop the below line

damacus commented on resources/firewall_nftables.rb:
```diff
@@ -0,0 +1,163 @@
+property :disabled, [true, false], default: false
+property :enabled, [true, false], default: true
+
+def whyrun_supported?
+  false
+end
```
We can just remove this entirely. 

```suggestion
```

damacus commented on resources/firewall_nftables.rb:
```diff
@@ -0,0 +1,163 @@
+end
+
+action :install do
+  return if disabled?(new_resource)
```
I don't really understand what we're doing here. We're letting the user use a resource but if they set it as disabled, then we don't run the code.
When would the user chose to use a resource but have it disabled? I've not come across this pattern before

a7b81a9086 replied:
The main reason this is there is probably, because I wanted to be compatible with the iptables provider. TBH, I also do not understand, why setting the resource to disabled disables all of the code completely. However, when fooling around with the firewall cookbook I found the following behaviour useful:
1. It is possible to create an invalid firewall ruleset. Generating a new ruleset before trying to restart the firewall service is important, because otherwise the cookbook fails because of a broken config and can thus never repair it.
1. Sometimes I like to read the rules before starting the firewall. Disabling the firewall but still generating the rules allows me to do that.
1. With firewalls, it can be annoying if you accidently enable it and lock yourself out. So maybe this was intended as a double safety that one is really sure to enable the firewall.

Points 1. and 2. of the above list unfortunately do not work with the current code, and it is questionable whether 3. is a valid point or not.

I can fix all this, or in fact, already have in another cookbook of mine. This resulted in much more modern and understandable code, using functions such as `edit_resource` which have been available in chef for some time.  Since the updates to the code were quite substantial: Is it okay for you if I just update this PR?

Stromweld replied:
It does seem odd to skip everything in the install. Again why use the resource then. I can see and use the disable firewall recipe to disable the service on systems in AWS as security groups are then used in place of OS firewall but maybe installed by default on image. Docker though requires firewall turned on for container networking to work.

The service though is then controlled with the service resource being added to wrapper cookbook. If you add service control to the resource I can see disabled property being used then. Not sure if that is pattern we’d want to use instead of pushing users to use the service resource along side this resource in their wrapper cookbook.

a7b81a9086 replied:
`firewall_rule` resources search for the `firewall['default']` resource, and add their rules there. There are however two ways to do this:
1. If `firewall['default']` has not been defined yet, define it automatically and add the rule there.
2. If `firewall['default']` has not been defined yet, print a warning, and do not add the rule.

Point 1. is way more flexible, because it offers the possibility to use `firewall_rule` resources in any place one would like, while with point 2. it is mandatory that `firewall['default']` was defined somewhere before in the run list. But with point 1 it is far from obvious to a user of the cookbook that defining a `firewall_rule` resource will automagically define a `firewall['default']` resource which will put that rule to action. In this case, an attribute which controls whether the `firewall['default']` is skipped or not comes in very handy.

Since the current implementation follows point 2, it is ok to skip the check.

bmhughes commented on libraries/helpers_nftables.rb:
```diff
@@ -0,0 +1,122 @@
+    module Nftables
+      include FirewallCookbook::Helpers
+
+      unless defined? CHAIN
```
We can use `||=` here instead of the nested conditional

a7b81a9086 replied:
done.

bmhughes commented on libraries/helpers_nftables.rb:
```diff
@@ -0,0 +1,122 @@
+        }.freeze
+      end
+
+      unless defined? TARGET
```
We can use `||=` here instead of the nested conditional

a7b81a9086 replied:
done.

bmhughes reviewed (changes requested):
Only a couple of style changes from me, otherwise looking good.

Stromweld commented:
couple of minor things for me and then if you could fix the stuff in the failing jobs. Otherwise nice work. Love seeing the modern resource in this CB finally.

damacus commented on README.md:
```diff
@@ -28,11 +28,13 @@ depends 'firewall'
 
 Tested on:
 
 - Ubuntu 16.04 with iptables, ufw
 - Debian 9 with iptables
+- Debian 11 with nftables
 - CentOS 6 with iptables
 - CentOS 7.1 with firewalld
 - Windows Server 2012r2 with Windows Advanced Firewall
```
Not on this PR, but definitely in another one if anyone is interested in updates there.

a7b81a9086 commented on kitchen.yml:
```diff
@@ -27,6 +27,9 @@ platforms:
   - name: debian-10
+  - name: debian-11
+    driver_config:
+      box: debian/bullseye64
```
Makes it easier for me, thanks for the hint!

a7b81a9086 commented:
Hi,

sorry for the mess, but I just added one more feature, support for multiple actions in one rule. I needed to add that, because otherwise logging packets would not have worked properly and I felt that this was kind of important to make this PR complete.

Anyway, I read all your comments and I believe I fixed all of them. The only exception is @Stromweld's request to fix errors in the pipeline. Yamllint yields no errors when I run it locally, but the github action fails with an error message that `docker pull` did not work. This seems to be some issue with the test itself, rather than my code, but have no idea how to fix this.

I will now stop to work on this PR, but again, if you have any comments or suggestions that need to be resolved before this can be merged, please tell me.

a7b81a9086 commented:
Hi,
with yamllint working, I reset the changes to the tests for this CB back to their original state, because it made all the tests fail.
I added `debian-11/nftables` to the github actions, and all tests pass.
FYI: `ufw` and `iptables` work on `debian-11`, `firewalld` fails, because it runs the ufw-provider instead of the firewalld-provider. Since these test have nothing to do with this PR I removed them again from github actions, as I did not add them on purpose.

The only thing that is blocking the merge now, are the changes requested by @bmhughes. @bmhughes: I think I have fixed them, so I took the liberty of requesting another review from you, so that you can mark it as resolved if you can spare the time.

kitchen-porter commented:
Released as: [5.0.0](https://github.com/sous-chefs/firewall/releases/tag/5.0.0)
